[testing] Add validation for concurrent cache in Conan

uilianries opened the pull request:
Hello!

This new validation introduces a test for concurrent. This initial validation started in this gist: https://gist.github.com/uilianries/e0438f6ac92f8444a1cb89c6faa3a4a7

So far, we found some conflicts:

- The test package uses the same folder when building, so parallel build will overwrite it
- Export source and Export will overwrite the same artifacts
- The result in Mac is a busy/locked file and folder 

/cc @AbrilRBS @memsharded 

Changelog: Omit
Docs: Omit

- [ ] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [ ] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).
- [ ] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

uilianries committed:
Add validation for concurrent cache in Conan

Signed-off-by: Uilian Ries <uilianries@gmail.com>

memsharded commented on test/functional/test_parallel_cache.py:
```diff
@@ -0,0 +1,55 @@
+from conan.test.assets.genconanfile import GenConanfile
+from conan.test.utils.tools import TestClient
+
+_parallel_options = [
```
The idea was to start first with concurrency for ``conan config install/install-pkg`` which hopefully would be a bit easier problem

uilianries replied:
Added the test `test_parallel_config` to validate conan config install

memsharded commented on test/functional/test_parallel_cache.py:
```diff
@@ -0,0 +1,55 @@
+    """
+    num_builds = len(_parallel_options)
+    test_client = TestClient()
+    test_client.run("new cmake_lib -d name=parallel -d version=0.1.0")
```
The concurrency over ``test_package`` would be a different problem. Lets focus first on the cache concurrency, the ``test_package`` can have other mechanisms, for example using ``build_folder_vars`` using the ``option`` will create independent folders.

uilianries replied:
Simplified the test to execute without test package. Plus, added a new one only to validate conan export. Please, see [86e341d](https://github.com/conan-io/conan/pull/18098/commits/86e341d5fc6e1954c573572b34253e1b629f628f)

uilianries committed:
Add test cases for config and export

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Remove assert for exceptions

It's not possible to guarantee the concurrent failure as it may depends
on the operate system as well. For instance, in MacOS the failure ocurrs
less than in Linux (based on tests executed in the CI)

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Add Winerror for Windows

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Add missing comma for error list

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Make sure there will be execeptions

Signed-off-by: Uilian Ries <uilianries@gmail.com>

uilianries committed:
Remove doubled variable

Signed-off-by: Uilian Ries <uilianries@gmail.com>
